pkg/controllers/terminating/v1alpha1/terminate.go

```diff
@@ -38,105 +38,62 @@ type Terminator struct {
+func (t *Terminator) cordonNode(ctx context.Context, node *v1.Node) error {
+	persisted := node.DeepCopy()
+	node.Spec.Unschedulable = true
+	if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+		return fmt.Errorf("patching node %s, %w", node.Name, err)
 	}
+	zap.S().Debugf("Cordoned node %s", node.Name)
 	return nil
 }
 
-// terminateNodes takes in a list of expired non-drained nodes and calls drain on them
-func (t *Terminator) terminateNodes(ctx context.Context) error {
-	// 1. Get draining nodes
-	draining, err := t.getLabeledNodes(ctx, provisioning.ProvisionerDrainingPhase)
+// drainNode evicts pods from the node and returns true when fully drained
+func (t *Terminator) drainNode(ctx context.Context, node *v1.Node) (bool, error) {
+	// 1. Get pods on node
+	pods, err := t.getPods(ctx, node)
 	if err != nil {
-		return fmt.Errorf("listing draining nodes, %w", err)
+		return false, fmt.Errorf("listing pods for node %s, %w", node.Name, err)
 	}
-	// 2. Drain nodes
-	drained := []*v1.Node{}
-	for _, node := range draining {
-		// TODO: Check if Node should be drained
-		// - Disrupts PDB
-		// - Pods owned by controller object
-		// - Pod on Node can't be rescheduled elsewhere
-
-		// 2a. Get pods on node
-		pods, err := t.getPods(ctx, node)
-		if err != nil {
-			return fmt.Errorf("listing pods for node %s, %w", node.Name, err)
-		}
-		// 2b. Evict pods on node
-		empty := true
-		for _, p := range pods {
-			if !pod.IsOwnedByDaemonSet(p) {
-				empty = false
-				if err := t.coreV1Client.Pods(p.Namespace).Evict(ctx, &v1beta1.Eviction{
-					ObjectMeta: metav1.ObjectMeta{
-						Name: p.Name,
-					},
-				}); err != nil {
-					zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
-				}
+	// 2. Evict pods on node
+	empty := true
+	for _, p := range pods {
+		if !pod.IsOwnedByDaemonSet(p) {
+			empty = false
+			if err := t.coreV1Client.Pods(p.Namespace).Evict(ctx, &v1beta1.Eviction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: p.Name,
+				},
+			}); err != nil {
+				zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
 			}
 		}
-		// 2c. If node is empty, add to list of nodes to delete
-		if empty {
-			drained = append(drained, node)
-		}
 	}
-	// 3. Delete empty nodes
-	if err := t.deleteNodes(ctx, drained); err != nil {
-		return fmt.Errorf("deleting %d nodes, %w", len(drained), err)
-	}
-	return nil
+	return empty, nil
 }
 
-// deleteNode uses a cloudprovider-specific delete to delete a set of nodes
-func (t *Terminator) deleteNodes(ctx context.Context, nodes []*v1.Node) error {
-	// 1. Delete node in cloudprovider's instanceprovider
-	if err := t.cloudProvider.Terminate(ctx, nodes); err != nil {
+// terminateNode terminates the node then removes the finalizer to delete the node
+func (t *Terminator) terminateNode(ctx context.Context, n *v1.Node) error {
+	if err := t.cloudProvider.Terminate(ctx, n); err != nil {
```

Comment:
I think we have to be super sure that `Terminate` won't return an error if the instance is truly gone, otherwise the finalizer might never be removed (that is, the code will never get to line 86 below)

Reply:
Good callout. I just thought about this. This cloudprovider's `Terminate()` can fail for a few reasons. The first two are acceptable, while the last one is known.
- The instance has termination protection
  - In this case, the termination API will return an error.
- EC2 Quota limits
  - In this case, we will exponentially back off and retry, as expected.

- The instance is not found
  - In this case, the node controller in CCM will attempt to clean up the resource (but be blocked by the finalizer), which will then call this function. This error will be a `InvalidInstanceID.NotFound` exception.
  - This is unlikely to happen, but could still happen, and was covered in the design doc. The next PR, which will only add the finalizer if the delete request is validated will lower the chances of this happening. This will make it so that only if Karpenter is uninstalled after the finalizer is added and instance is terminated, but the finalizer is not removed (which happens back-to-back, so is super super unlikely), and then remains uninstalled for ~1 hour (time for terminated ec2 instances to get cleaned up in console), then is reinstalled, that we will run into this problem.

Reply:
I wonder if it would just be simplest to treat InvalidInstanceId.NotFound as a non-error?

Reply:
You're right. I added it in.